### DISCUSSION TITLE
Pinentry no longer grabs focus

### DISCUSSION THREAD

**Discussion by viceice** [git-for-windows/git#3789] [General]
Since one of the v2.35+ git updates the intgrasted pinentry popup for gpg signing no longer grabs the focus on windows 11.

Any ideas why this could happen?

Can it be related to [windows powertoys](https://github.com/microsoft/PowerToys) ?

**Comment by viceice** [comment]
Initially thought it was a vscode update
- https://github.com/microsoft/vscode/issues/147371

**Comment by dscho** [comment]
Hmm. Neither https://github.com/git-for-windows/git/releases/v2.35.2.windows.1 nor https://github.com/git-for-windows/git/releases/v2.35.3.windows.1 updates GPG... Could you verify that your `gpg --version` says v2.2.29?

**Reply by viceice** [reply]
yes
```sh
$ gpg --version
gpg (GnuPG) 2.2.29-unknown
```
I'm not sure when it started to occure. sorry.

**Reply by dscho** [reply]
Just for the record, the pinentry window still grabs the focus for me...

**Reply by viceice** [reply]
ok, so it must be a windows 11 or power toys issue. 🤔

i recently switched from a dual monitor to a single ultra wide. so i use fancy zones to align my windows.

I'll try again without power toys to verify.

**Reply by viceice** [reply]
Ok, on a notebook without powertoys it seems to work normally. Also after restart powertoys / gpg-agent evething works normally now 🤔

**Reply by viceice** [reply]
Now happened again using vscode on wsl remote

I've a gpg wrapper to call the git4windows gpg executable

```sh
> cat ~/bin/gpg
#!/bin/sh

exec '/mnt/c/Program Files/Git/usr/bin/gpg.exe' "$@"
```

**Reply by viceice** [reply]
ok, when starting the gpg agent from a normal windows console it works, but when the agent is started from wsl it doesn't work

**Reply by viceice** [reply]
so my workaround is now to start the agent via windows task on logon

**Reply by dscho** [reply]
Ah, that would make sense. I wonder whether there is any trick to get GUI windows started via WSL to be opened in the foreground. (FWIW I am also running on Windows 11.)

**Reply by viceice** [reply]
didn't found something yet. starting via scheduled task only helps for some time, it seems the agent is stopping automatically after some time?

**Reply by viceice** [reply]
Strange, i now see same behavior on windows server 2016 when commiting from vscode 😕

**Comment by viceice** [comment]
Found this: https://dev.gnupg.org/T4123

**Reply by dscho** [reply]
Huh. I thought we were up to date. Looking at https://github.com/gpg/gnupg/tags suggests that v2.2.29 is current, which is exactly the GNU Privacy Guard version we ship.

But then I had a look at https://gnupg.org/ and it claims that v2.2.40 is current?!?

**Reply by viceice** [reply]
yes, it seems they no longer mirror the git tags from https://git.gnupg.org/cgi-bin/gitweb.cgi?p=gnupg.git

**Reply by dscho** [reply]
I think I have addressed the problem with the missing ticket here: https://github.com/git-for-windows/track-website-changes/commit/44ff01d969041e889b240605621db807e80ac27b

**Reply by viceice** [reply]
i'll test again on next release 🤞

**Reply by dscho** [reply]
Testing on releases is a bit late. I put in quite an effort to provide [snapshots](https://wingit.blob.core.windows.net/files/index.html) and pre-releases so that users like you can help prevent breakages in releases. I rely on you to not wait for releases to test.

**Reply by dscho** [reply]
@viceice please download https://github.com/git-for-windows/MSYS2-packages/suites/9918749713/artifacts/479701703, unpack the `.tar.zst` package that is contained within, and then extract the `/usr/bin/` and `/usr/lib/` files from that tar file into a PortableGit, overwriting the existing files, then test.

**Reply by viceice** [reply]
ok, can try on Monday

**Reply by dscho** [reply]
@viceice please do. I spent quite a bit of effort on addressing the issue you reported.

**Reply by viceice** [reply]
tested, window is on top but still not focused. 😕

**Reply by dscho** [reply]
@viceice and this is with a pinentry version that is supposed to fix the issue?

**Reply by viceice** [reply]
🤷‍♂️ At least the version of the package above isn't working. Not sure which pinentry it includes, but the currently working from gpg4windows seems to be `pinentry-qt`

![image](https://user-images.githubusercontent.com/1798109/208471668-c3fa5f08-69f4-4783-8e87-467b1a271d51.png)

**Reply by viceice** [reply]
and the non-working ui
![image](https://user-images.githubusercontent.com/1798109/208472015-195ca02b-19c9-4363-8ed7-fc5438e52dcc.png)

**Comment by viceice** [comment]
my workaround now is to install gpg4win and use that gpg for git. there's a newer pinentry wich successfully comes to front
